M2/Macaulay2/packages/Cremona.m2

```diff
@@ -583,18 +650,30 @@ RationalMap ^ ZZ := (Phi,j) -> (
          if (target Phi#"map" === source Phi#"map") then return rationalMap(target Phi#"map") else error "expected non-zero integer";
    );
    if j < 0 then (
-         Psi:=inverseMap(Phi,MathMode=>true,Verbose=>false);
+         Psi:=inverse Phi;
          return(Psi^(-j))
    );
    Psi2:=Phi; for i from 1 to j-1 do Psi2 = Psi2 * Phi; 
    return Psi2;
 );
 
-inverse (RationalMap) := (Phi) -> Phi^(-1);
+inverse (RationalMap,Option) := (Phi,opt) -> (
+   o := toList opt;
+   if not(#o == 2 and first o === MathMode) then error "MathMode is the only available option for inverse(RationalMap)";
+   if not instance(last o,Boolean) then error "option MathMode accepts true or false";
+   if Phi#"inverseRationalMap" =!= null then return Phi#"inverseRationalMap";
+   if last o then return inverseMap(Phi,MathMode=>true,Verbose=>false);
+   Psi := inverseMap Phi;
+   if not isInverseMapFast(Phi,Psi) then error "failed to get inverse map; make sure the input map is birational";
+   if Phi#"inverseRationalMap" === null and Psi#"inverseRationalMap" === null then forceInverseMap(Phi,Psi); 
+   Psi
+);
+
+inverse (RationalMap) := (Phi) -> inverse(Phi,MathMode=>false);
 
 RationalMap ! := (Phi) -> (
      toMap Phi;
-     try Phi^(-1);
+     try inverse Phi;
```

Comment:
Using "try" is unreliable, because if it may catch an error of a type you didn't expect, which you would want to find out about.  Why even run this code, if the result is not needed?

Reply:
Ok, but this is really a very secondary function (mostly used for testing). I would leave it unchanged if possible.

Reply:
This function is only useful for some simple maps where you can calculate everything easily. It returns the same rational map, i.e. the same mutable hash table, but first fills almost all the entries (such as "projectiveDegrees", "isBirational", etc). After this run, the method "describe(RationalMap)" yields a larger output.
Example:
`i1 : needsPackage "Cremona";

i2 : f = rationalMap map specialCremonaTransformation 5;

o2 : RationalMap (quadratic rational map from PP^5 to PP^5)

i3 : describe f

o3 = rational map defined by forms of degree 2
     source variety: PP^5
     target variety: PP^5
     coefficient ring: QQ

i4 : describe f!

o4 = rational map defined by forms of degree 2
     source variety: PP^5
     target variety: PP^5
     dominance: true
     birationality: true (the inverse map is already calculated)
     projective degrees: {1, 2, 4, 4, 2, 1}
     number of minimal representatives: 1
     dimension base locus: 2
     degree base locus: 4
     coefficient ring: QQ
`
The user should perform the functions he needs, not all of them in the order I choose. So I consider this function quite irrelevant.

Reply:
What if you just remove "try"?  Won't it still work?

Reply:
In that case, it would give an error when the map `Phi` is not birational. One could replace `try inverse Phi` with `if isBirational Phi then inverse Phi`, but this test requires non trivial computation (and useless if in the case of birationality one also want the inverse map).

Reply:
If `inverse Phi` fails more quickly than `isBirational Phi` does when `Phi` is not birational, why not include the code that fails at the start of `isBirational`?  Then `isBirational Phi` will return `false` more quickly.

Reply:
This could be done, but I have some concerns about it. I already use this approach. If `inverse Phi` succeeds then `isBirational Phi` will return `true` quickly; if `degree Phi` returns a value different from 1 then `isBirational Phi` will return `false` quickly and `inverse Phi` will give an error quickly, etc. But I don't trust that a particular code fails only in the case the answer doesn't exist. In the future, the code of `inverse(RationalMap)` could be rewritten keeping track of where it fails (if it does).

Reply:
Okay.